File: LeetCode/Easy/0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countNodes(root *TreeNode) int {
    cnt := 0
    visitNode(&cnt, root)
    return cnt
}

func visitNode(cnt *int, cur *TreeNode) {
    if (cur != nil) {
        *cnt += 1
        visitNode(cnt, cur.Left)
        visitNode(cnt, cur.Right)
    }
    return
}

File: LeetCode/Easy/0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftDepth = getDepth(root.left);
        int rightDepth = getDepth(root.right);

        if (leftDepth == rightDepth) {
            return (1 << leftDepth) - 1 + countNodes(root.right) + 1;
        }

        return (1 << rightDepth) - 1 + countNodes(root.left) + 1;
    }

    public int getDepth(TreeNode node) {
        if (node == null) {
            return 0;
        }
        return 1 + getDepth(node.left);
    }
}